feat: Release notes for Codacy Self-hosted 12.0.0 DOCS-561 (#1767)
codacy/docs: docs/release-notes/self-hosted/self-hosted-v12.0.0.md

---
rss_title: Codacy release notes RSS feed
rss_href: /feed_rss_created.xml
description: Release notes for Codacy Self-hosted v12.0.0.
codacy_tools_version_old: https://github.com/codacy/codacy-tools-sh/releases/tag/sh-1.1.6

> codacy_tools_version_new: https://github.com/codacy/codacy-tools-sh/releases/tag/sh-1.2.1

Reviewer: This is a minor detail but it can cause some confusion if we're debugging something later on. :nerd_face: 

The latest commit in the release branch is already using version 1.2.2 of the codacy-tools-sh (that's what brought in a couple more tool updates):

```suggestion
codacy_tools_version_new: https://github.com/codacy/codacy-tools-sh/releases/tag/sh-1.2.2
```

## Upgrading Codacy Self-hosted

Follow the steps below to upgrade to Codacy Self-hosted v12.0.0:

1.  Check the [release notes for all Codacy Self-hosted versions](../index.md#self-hosted) **between your current version and the most recent version** for breaking changes and follow the instructions provided <span class="skip-vale">carefully</span>.

1.  Follow the instructions to [upgrade your Codacy Self-hosted instance](https://docs.codacy.com/v12.0/chart/maintenance/upgrade/).

1.  Update your Codacy command-line tools to the following versions:<!--TODO Update CLI tool versions-->

    -   [Codacy Analysis CLI MAJOR.MINOR.PATCH](https://github.com/codacy/codacy-analysis-cli/releases/tag/MAJOR.MINOR.PATCH)
    -   [Codacy Coverage Reporter MAJOR.MINOR.PATCH](https://github.com/codacy/codacy-coverage-reporter/releases/tag/MAJOR.MINOR.PATCH)

## Breaking changes

-   If you're updating Kubernetes to v1.25, upgrade to Codacy Self-hosted v12.0.0 before doing so, as some Kubernetes endpoints have been deprecated and may lead to service outage if used with earlier versions of Codacy Self-hosted.

> -   If using the monitoring functionality [Grafana + Prometheus + Loki](https://docs.codacy.com/chart/configuration/monitoring/#setting-up-monitoring-using-grafana-prometheus-and-loki), ensure Pod Security Policies are disabled for Loki and Promtail by following the relevant update instructions ([Loki](https://docs.codacy.com/chart/configuration/monitoring/#2-installing-loki), [Promtail](https://docs.codacy.com/chart/configuration/monitoring/#3-installing-promtail)) after upgrading to Codacy Self-hosted v12.0.0 and before updating Kubernetes to v1.25.

Reviewer: @nicklem Let's add the terms "Grafana" and "Promtail" to the Vale accept list to clean up the errors.

> -   If using the monitoring functionality [Grafana + Prometheus + Loki](https://docs.codacy.com/chart/configuration/monitoring/#setting-up-monitoring-using-grafana-prometheus-and-loki), ensure Pod Security Policies are disabled for Loki and Promtail by following the relevant update instructions ([Loki](https://docs.codacy.com/chart/configuration/monitoring/#2-installing-loki), [Promtail](https://docs.codacy.com/chart/configuration/monitoring/#3-installing-promtail)) after upgrading to Codacy Self-hosted v12.0.0 and before updating Kubernetes to v1.25.

Reviewer 2: @prcr I updated the instructions to highlight the execution order.

I had to use HTML to force an ordered list. I'm not sure it's worth the effort, though. We may want to revert to an unordered one. Thoughts?

> -   If using the monitoring functionality [Grafana + Prometheus + Loki](https://docs.codacy.com/chart/configuration/monitoring/#setting-up-monitoring-using-grafana-prometheus-and-loki), ensure Pod Security Policies are disabled for Loki and Promtail by following the relevant update instructions ([Loki](https://docs.codacy.com/chart/configuration/monitoring/#2-installing-loki), [Promtail](https://docs.codacy.com/chart/configuration/monitoring/#3-installing-promtail)) after upgrading to Codacy Self-hosted v12.0.0 and before updating Kubernetes to v1.25.

Reviewer: The "scheduler" instructions below should also use an ordered list like this. :slightly_smiling_face:

> -   If using the monitoring functionality [Grafana + Prometheus + Loki](https://docs.codacy.com/chart/configuration/monitoring/#setting-up-monitoring-using-grafana-prometheus-and-loki), ensure Pod Security Policies are disabled for Loki and Promtail by following the relevant update instructions ([Loki](https://docs.codacy.com/chart/configuration/monitoring/#2-installing-loki), [Promtail](https://docs.codacy.com/chart/configuration/monitoring/#3-installing-promtail)) after upgrading to Codacy Self-hosted v12.0.0 and before updating Kubernetes to v1.25.

Reviewer 2: Correct, it does work but renders as letters. I just didn't like it that much 😅 

I'm reverting to an ordered list. I'll apply the same to the scheduler instructions as suggested.

-   If you have set `scheduler.enable = "true"` in `values.yaml`, before migrating to Codacy Self-hosted v12.0.0:

    -   Remove scheduler settings from `values.yaml`
    -   In `worker-manager`, ConfigMap set CONFIG_FORCE_codacy_kubernetes_scheduler_name: default-scheduler 
    -   Restart `worker-manager` deployment

>     -   Remove any pending worker Pods

Reviewer: These instructions read like a draft. Is there any chance we can make them more explicit for customers?

For example:

- Remove scheduler settings from `values.yaml` :arrow_right: perhaps another way of saying this could be to remove all `scheduler.*` entries from the file?
- In `worker-manager`, ConfigMap set CONFIG_FORCE_codacy_kubernetes_scheduler_name: default-scheduler :arrow_right: what does this actually mean? What is `worker-manager`? Is it a pod? The "ConfigMap set [...]" looks like it could be a command... If it is, how do users run it?
- Restart `worker-manager` deployment :arrow_right: Can we provide a command for doing this?
- Remove any pending worker Pods :arrow_right: How do users check if there are "pending worker pods"? Which pods are those? Again, is there a command we can provide to remove those pods?

@tfrgoncalves Perhaps you can help clarify some of these questions?

Reviewer 3: 1. Make sense, indeed. 
2. [Here](https://github.com/codacy/docs/pull/1767#discussion_r1239966685), I wrote some other options for us to write it programatically, we just need to choose one. In this case the ConfigMap is called `worker-manager`, and is used by a `worker-manager` deployment.
3. `kubectl rollout restart deployment/worker-manager -n codacy`
4. `kubectl get pods -l app=worker --field-selector=status.phase=Pending -n codacy` if result is empty no pods are pending.

Let me know if any of these need further clarification.

Reviewer 2: @tfrgoncalves I updated the [breaking changes instructions](https://feat-release-notes-self-hosted-12-0--docs-codacy.netlify.app/release-notes/self-hosted/self-hosted-v12.0.0/#breaking-changes).

* Please take a look and let us know if everything is correct.
* We'll also need the final command to delete pending worker pods. I understand that this is a destructive action, and added a warning before it.

Thanks!

1.  Follow the instructions to [upgrade your Codacy Self-hosted instance](https://docs.codacy.com/v12.0/chart/maintenance/upgrade/).

1.  Update your Codacy command-line tools to the following versions:<!--TODO Update CLI tool versions-->

    -   [Codacy Analysis CLI MAJOR.MINOR.PATCH](https://github.com/codacy/codacy-analysis-cli/releases/tag/MAJOR.MINOR.PATCH)
    -   [Codacy Coverage Reporter MAJOR.MINOR.PATCH](https://github.com/codacy/codacy-coverage-reporter/releases/tag/MAJOR.MINOR.PATCH)

## Breaking changes

If you're updating Kubernetes to v1.25, upgrade to Codacy Self-hosted v12.0.0 before doing so, as some Kubernetes endpoints have been deprecated and may lead to service outage if used with Codacy Self-hosted v11.0.0.

If using the monitoring functionality [Grafana + Prometheus + Loki](https://docs.codacy.com/chart/configuration/monitoring/#setting-up-monitoring-using-grafana-prometheus-and-loki), ensure Pod Security Policies are disabled for Loki and Promtail by running the following commands after upgrading to Codacy Self-hosted v12.0.0 but before updating Kubernetes to v1.25:

> 
>  ```
>  command here
>  ```

Reviewer 2: ```suggestion
```

---
rss_title: Codacy release notes RSS feed
rss_href: /feed_rss_created.xml
description: Release notes for Codacy Self-hosted v12.0.0.
codacy_tools_version_old: https://github.com/codacy/codacy-tools-sh/releases/tag/sh-1.1.6
codacy_tools_version_new: https://github.com/codacy/codacy-tools-sh/releases/tag/sh-1.2.1
---

# Self-hosted v12.0.0

These release notes are for [Codacy Self-hosted v12.0.0](https://github.com/codacy/chart/releases/tag/12.0.0){: target="_blank"}, released on June 21, 2023.<!-- TODO Update release date -->

📢 [Visit the Codacy roadmap](https://roadmap.codacy.com) and <span class="skip-vale">let us know</span> your feedback on both new and planned product updates!

<!--TODO Check these issues manually
Jira issues without release notes

> -   https://codacy.atlassian.net/browse/TS-110 TODO review

Reviewer: From what I could understand, this was an internal refactoring with no impact for customers. I set the release note field to "Don't publish" and we can remove this line.

Jira issues with disabled release notes
-   https://codacy.atlassian.net/browse/TS-369 TODO Notify Paulo that we're ignoring this one: https://codacy.atlassian.net/browse/TS-369?focusedCommentId=56541

> -   https://codacy.atlassian.net/browse/TS-356 TODO we may want to notify this: ESlint: install plugin vue-scoped-css

Reviewer 2: This release note is flagged as "don't publish".
Any reason for not mentioning adding the `vue-scoped-css` plugin? I suggest including this in the release notes.

Reviewer: I'm not sure why the field was set to "Don't publish", we should definitely include this plugin in the list as well. :+1:

-->

## Upgrading Codacy Self-hosted

Follow the steps below to upgrade to Codacy Self-hosted v12.0.0:

1.  Check the [release notes for all Codacy Self-hosted versions](../index.md#self-hosted) **between your current version and the most recent version** for breaking changes and follow the instructions provided <span class="skip-vale">carefully</span>.

1.  Follow the instructions to [upgrade your Codacy Self-hosted instance](https://docs.codacy.com/v12.0/chart/maintenance/upgrade/).

1.  Update your Codacy command-line tools to the following versions:<!--TODO Update CLI tool versions-->

    -   [Codacy Analysis CLI MAJOR.MINOR.PATCH](https://github.com/codacy/codacy-analysis-cli/releases/tag/MAJOR.MINOR.PATCH)
    -   [Codacy Coverage Reporter MAJOR.MINOR.PATCH](https://github.com/codacy/codacy-coverage-reporter/releases/tag/MAJOR.MINOR.PATCH)

> ## Breaking changes

Reviewer: This version seems to introduce multiple breaking changes.

To make it more explicit that there are different breaking changes that customers must address, we can follow a format [similar to this example](https://docs.codacy.com/release-notes/self-hosted/self-hosted-v6.0.0/#breaking-changes). Basically, it involves listing each breaking change on its dedicated bullet item so users can more easily tell them apart.

If you're updating Kubernetes to v1.25, upgrade to Codacy Self-hosted v12.0.0 before doing so, as some Kubernetes endpoints have been deprecated and may lead to service outage if used with Codacy Self-hosted v11.0.0.

> If using the monitoring functionality [Grafana + Prometheus + Loki](https://docs.codacy.com/chart/configuration/monitoring/#setting-up-monitoring-using-grafana-prometheus-and-loki), ensure Pod Security Policies are disabled for Loki and Promtail by running the following commands after upgrading to Codacy Self-hosted v12.0.0 but before updating Kubernetes to v1.25:

Reviewer 2: @tfrgoncalves

We may need to add this command to the [relevant upgrade page](https://docs.codacy.com/chart/maintenance/upgrade/).

Could you confirm? Would updating point 4 of that page be enough?

Reviewer 3: What happens is that, if a user has the Grafana + ... installation in use and ran those helm commands we provide they need to disable those policies by running the helm update for them before upgrading to kubernetes 1.25. So, it is a "before upgrading to Kubernetes X" requirement. It may be more clear for users to have an approach similar to 'In particular, Codacy Self-hosted v5.0.0...' by the beginning of the page. @lolgab does it also make sense for you?

Reviewer 2: Thanks @tfrgoncalves! So, to clarify (keep in mind I have zero context on these procedures, so I'm guesstimating step 2):

1. [Upgrade Codacy](https://docs.codacy.com/chart/maintenance/upgrade/) (you mention "preferably", but it's best to give clear steps here)
2. Run a `helm update` on Loki and Promtail, essentially using the instructions we already provide ([Loki](https://docs.codacy.com/chart/configuration/monitoring/#2-installing-loki), [Promtail](https://docs.codacy.com/chart/configuration/monitoring/#3-installing-promtail))
  * When updating, customize [values-loki.yaml](https://docs.codacy.com/chart/values-files/values-loki.yaml) and [values-promtail.yaml](https://docs.codacy.com/chart/values-files/values-promtail.yaml) to include the option `podSecurityPolicy: enabled: false` (formatted as yaml)
  * Note: if the above holds true, I suggest updating those files for the v12 docs
3. Finally, update to Kubernetes 1.25

Makes sense?

Reviewer 3: All making sense. Upgrading Codacy to this version must always happen before upgrading Kubernetes to 1.25 indeed, not only "preferably".

Reviewer 2: Thanks!

While updating `values-loki.yaml` and `values-promtail.yaml`, I noticed that there's a line already setting the PSP value to false. I suspect that the syntax of the option `podSecurityPolicy` that I mentioned above may be incorrect, since it was suggested by ChatGPT. The one below is likely the correct one. I'd simply add a comment to the relevant line to clarify when not to change its value:

```
rbac:
  pspEnabled: false # must be set to `false` with Kubernetes 1.25 and higher
```

I'd then simplify the note to read something like this, plus an admonition similar to 'In particular, Codacy Self-hosted v5.0.0...' on the upgrade page, as you suggest.

```suggestion
If using the monitoring functionality [Grafana + Prometheus + Loki](https://docs.codacy.com/chart/configuration/monitoring/#setting-up-monitoring-using-grafana-prometheus-and-loki), ensure Pod Security Policies are disabled for Loki and Promtail by following the relevant update instructions ([Loki](https://docs.codacy.com/chart/configuration/monitoring/#2-installing-loki), [Promtail](https://docs.codacy.com/chart/configuration/monitoring/#3-installing-promtail)) after upgrading to Codacy Self-hosted v12.0.0 but before updating Kubernetes to v1.25.
```

Reviewer 3: The idea is for users to run helm upgrade against our most recent x-values.yaml, as seen in these two changes ([here](https://github.com/codacy/chart/commit/e246a14770aad096d1d1d8c85db018396e2ec079) and [here](https://github.com/codacy/chart/commit/c3423b632ffc2ceb165074d576c09ef9f37c6aec)) I made some time ago.

There is also the case users have not set up the [Grafana + Prometheus + Loki] installation in accordance to our values, which may lead to issues on those deployments, but that is out of our concern to solve.

Reviewer 3: Had overseen the `podSecurityPolicy: enabled: false` error part, sorry.

Reviewer 2: Thanks for clarifying. I'll update this line and avoid too much detail on the yaml files.

Reviewer 3: Long story short, they just need to rerun the helm upgrade if they used our values.yaml files or make sure their deployments disable rbac otherwise, before the upgrade to kubernetes 1.25.

If you have set `scheduler.enable = "true"` in `values.yaml`, before migrating to Codacy Self-hosted v12.0.0:

-   Remove scheduler settings from `values.yaml`
-   In `worker-manager`, ConfigMap set CONFIG_FORCE_codacy_kubernetes_scheduler_name: default-scheduler 
-   Restart `worker-manager` deployment

> -   Remove any pending worker Pods

Reviewer 2: @tfrgoncalves A couple notes:

`scheduler.enable = "true"`
I'm checking [`values.yaml`](https://github.com/codacy/chart/blob/master/codacy/values.yaml) and see lots of `enabled` keys but no `enable`. Just double-checking to confirm there's no typo here.

`ConfigMap set CONFIG_FORCE_codacy_kubernetes_scheduler_name: default-scheduler`
Can you provide more context on how this is run and provide an example? Just so I know how to format these commands (is it from a command line, config file, ...)

Reviewer 3: Rechecking always makes sense. We have [this](https://github.com/codacy/worker-manager-sh/blob/3515fe90883b7c21a1b34c6fa98c8fe93af9aea7/conf/reference.conf#L449) on the documentation and [here](https://github.com/codacy/worker-manager-sh/pull/9/files#diff-5f9d82490629d7c4b8aab497e5a44971aee6703afccf21ca665498a8cd8c5acaL55) Lorenzo is removing it.

Regarding the ConfigMap, I am proposing a 'manual' edition of the Configmap. We could provide a kubectl command as I mentioned before, but that may pose some dangers. I will leave all existing options below.

Assuming codacy is installed in namespace 'codacy' as usual:
Manual approach: `kubectl edit configmap worker-manager --namespace codacy`, change said value to 'default-scheduler' and save.

Automatic approach: `kubectl patch configmap/worker-manager --namespace codacy --type merge --patch '{"data":{"CONFIG_FORCE_codacy_kubernetes_scheduler_name":"default-scheduler"}}'`

Shorter automatic approach: `kubectl patch configmap/worker-manager -n codacy --type merge -p '{"data":{"CONFIG_FORCE_codacy_kubernetes_scheduler_name":"default-scheduler"}}'`

We may leave the last option with a note to contact support in case of need.

## Product enhancements

-   Added support for [Kubernetes 1.25.\*](https://docs.codacy.com/v12.0/chart/requirements/#kubernetes-or-microk8s-cluster-setup) (REL-1306)
-   Added support for the ESLint plugin [<span class="skip-vale">eslint-plugin-rxjs</span>](https://www.npmjs.com/package/eslint-plugin-rxjs). (TS-346)

> -   Added support for the ESLint plugins [<span class="skip-vale">eslint-plugin-markdown</span>](https://www.npmjs.com/package/eslint-plugin-markdown) and [<span class="skip-vale">eslint-plugin-yml</span>](https://www.npmjs.com/package/eslint-plugin-yml). (TS-121)

Reviewer: When we add multiple plugins for the same tool, we can manually group them together like [in this example](https://github.com/codacy/docs/blob/master/docs/release-notes/cloud/cloud-2023-05.md?plain=1#L18). It's just a tidier way to present the information to the users.

> ## Tool versions

Reviewer: We should re-run the release notes script again taking into account the latest commits pushed to the release branch on the codacy/chart repository so we're sure that we got all the updated tools (see [this message](https://codacy.slack.com/archives/CSSSHAK9N/p1687786409275869?thread_ts=1687166816.965619&cid=CSSSHAK9N) from @lolgab).

Bugs and Community Issues:
-   https://codacy.atlassian.net/browse/TS-374
-   https://codacy.atlassian.net/browse/TS-369
-   https://codacy.atlassian.net/browse/TS-356
-   https://codacy.atlassian.net/browse/TS-352
-   https://codacy.atlassian.net/browse/IO-624
-   https://codacy.atlassian.net/browse/IO-541
-->

## Upgrading Codacy Self-hosted

Follow the steps below to upgrade to Codacy Self-hosted v12.0.0:

1.  Check the [release notes for all Codacy Self-hosted versions](../index.md#self-hosted) **between your current version and the most recent version** for breaking changes and follow the instructions provided <span class="skip-vale">carefully</span>.

1.  Follow the instructions to [upgrade your Codacy Self-hosted instance](https://docs.codacy.com/v12.0/chart/maintenance/upgrade/).

1.  Update your Codacy command-line tools to the following versions:<!--TODO Update CLI tool versions-->

    -   [Codacy Analysis CLI MAJOR.MINOR.PATCH](https://github.com/codacy/codacy-analysis-cli/releases/tag/MAJOR.MINOR.PATCH)
    -   [Codacy Coverage Reporter MAJOR.MINOR.PATCH](https://github.com/codacy/codacy-coverage-reporter/releases/tag/MAJOR.MINOR.PATCH)

## Product enhancements

> ## Bug fixes

Author: These are new ESLint plugins, not bug fixes.
We need a better title.

Reviewer 2: Agree. Moving these to product enhancements.

---

# Self-hosted v12.0.0

These release notes are for [Codacy Self-hosted v12.0.0](https://github.com/codacy/chart/releases/tag/12.0.0){: target="_blank"}, released on June 21, 2023.<!-- TODO Update release date -->

📢 [Visit the Codacy roadmap](https://roadmap.codacy.com) and <span class="skip-vale">let us know</span> your feedback on both new and planned product updates!

## Upgrading Codacy Self-hosted

Follow the steps below to upgrade to Codacy Self-hosted v12.0.0:

1.  Check the [release notes for all Codacy Self-hosted versions](../index.md#self-hosted) **between your current version and the most recent version** for breaking changes and follow the instructions provided <span class="skip-vale">carefully</span>.

1.  Follow the instructions to [upgrade your Codacy Self-hosted instance](https://docs.codacy.com/v12.0/chart/maintenance/upgrade/).

1.  Update your Codacy command-line tools to the following versions:

    -   [Codacy Analysis CLI 7.8.3](https://github.com/codacy/codacy-analysis-cli/releases/tag/7.8.3)
    -   [Codacy Coverage Reporter 13.13.1](https://github.com/codacy/codacy-coverage-reporter/releases/tag/13.13.1)

## Breaking changes

> - If you have set `scheduler.enable = "true"` in `values.yaml`, do the following.

Reviewer: It doesn't hurt to explicitly state that this part must also be performed before actually updating Kubernetes (either 1.24 or 1.25, it doesn't matter):

```suggestion
-   If you have set `scheduler.enable = "true"` in `values.yaml`, do the following before updating Kubernetes.
```

> -   If you have set `scheduler.enable = "true"` in `values.yaml`, do the following before updating Kubernetes.

Reviewer 2: ```suggestion
-   If you have set `scheduler.enable: "true"` in `values.yaml`, do the following before updating Kubernetes.
```

> Codacy Self-hosted v12 supports Kubernetes v1.24 and v1.25. If you're upgrading your Kubernetes installation, make sure to do the following in the recommended order to avoid service outages:

Reviewer 2: Removing v1.24, as we now flag SH v11 as [already compatible with Kubernetes v1.24](https://docs.codacy.com/chart/requirements/#kubernetes-or-microk8s-cluster-setup).

```suggestion
Codacy Self-hosted v12 supports Kubernetes v1.25. If you're upgrading your Kubernetes installation, make sure to do the following in the recommended order to avoid service outages:
```

> -   If you have set `scheduler.enable: "true"` in `values.yaml`, do the following before updating Kubernetes.
> 
>     Assuming you're using the default `codacy` namespace:
> 
>     1.  Remove all `scheduler.*` settings from `values.yaml`

Reviewer 2: @tfrgoncalves Can you validate? See [Ranga's comment](https://github.com/codacy/docs/pull/1767#discussion_r1267027931) for more context.

cc: @Ngwerume

```suggestion
-   If you have set `scheduler.enable: "true"` in `values-production.yaml`, do the following before updating Kubernetes.

    Assuming you're using the default `codacy` namespace:

    1.  Remove all `scheduler.*` settings from `values-production.yaml`
```

Reviewer 3: Considering user overrides should be done in values-production.yaml and values-microk8s.yaml both should probably be mentioned. @lolgab @Ngwerume, do you agree?

Author: As it's [written on the installation docs](https://github.com/codacy/chart/blob/082dab58d173ca5824a2d9994f7d979b9a414eb3/docs/index.md#L80), `values-microk8s.yaml` can be used instead of `values-production.yaml` so you might want to add a notice saying that the changes are to be applied to `values-microk8s.yaml` if your are using MicroK8s.

Reviewer 2: Thanks @lolgab ! Here the updated suggestion:

cc: @Ngwerume @tfrgoncalves 

```suggestion
-   If you have set `scheduler.enable: "true"` in `values-production.yaml` (or `values-microk8s.yaml` if using MicroK8s), do the following before updating Kubernetes.

    Assuming you're using the default `codacy` namespace:

    1.  Remove all `scheduler.*` settings from `values-production.yaml` (or `values-microk8s.yaml` if using MicroK8s)
```

>     1.  Update the ConfigMap of `worker-manager`:
> 
>         ```bash
>         kubectl patch configmap/worker-manager --namespace codacy --type merge --patch '{"data":{"CONFIG_FORCE_codacy_kubernetes_scheduler_enable": "false","CONFIG_FORCE_codacy_kubernetes_scheduler_name":"default-scheduler"}}'
>         ```
> 
>     1.  Restart the `worker-manager` Deployment:
> 
>         ```bash
>         kubectl rollout restart deployment/worker-manager -n codacy
>         ```
> 

Reviewer 4: Correct resource name.
@tfrgoncalves Please validate. 🙏 

```suggestion
    1.  Update the ConfigMap of `codacy-worker-manager`:

        ```bash
        kubectl patch configmap/codacy-worker-manager --namespace codacy --type merge --patch '{"data":{"CONFIG_FORCE_codacy_kubernetes_scheduler_enable": "false","CONFIG_FORCE_codacy_kubernetes_scheduler_name":"default-scheduler"}}'
        ```

    1.  Restart the `codacy-worker-manager` Deployment:

        ```bash
        kubectl rollout restart deployment/codacy-worker-manager -n codacy
        ```

```

Reviewer 4: @tfrgoncalves Please validate. 🙏

Reviewer 3: LGTM. Thank you @Ngwerume for the find and @claudiacarpinteiro for the fix.

    1.  List any pending `worker` Pods:

        ```bash
        kubectl get pods -l app=worker --field-selector=status.phase=Pending -n codacy
        ```

    1.  If there are any pending `worker` Pods, remove them:

        !!! warning
            This is a destructive action. Make sure you understand the consequences.

        ```bash
        kubectl delete pods -l app=worker --field-selector=status.phase=Pending -n codacy
        ```

-   If you're upgrading your Kubernetes installation to v1.25, make sure to do the following in the recommended order to avoid service outages:

    1.  Upgrade to Codacy Self-hosted v12.0.0

>     1.  If using the monitoring functionality [Grafana + Prometheus + Loki](https://docs.codacy.com/v12.    0/chart/configuration/monitoring/#setting-up-monitoring-using-grafana-prometheus-and-loki), update Loki and Promtail by following the     latest respective installation instructions ([Loki](https://docs.codacy.com/v12.0/chart/configuration/monitoring/#2-installing-loki), [Promtail](https://docs.codacy.com/v12.0/chart/configuration/monitoring/#3-installing-promtail)).

Reviewer: It seems like there are some spurious spaces on this line:

```suggestion
    1.  If using the monitoring functionality [Grafana + Prometheus + Loki](https://docs.codacy.com/v12.0/chart/configuration/monitoring/#setting-up-monitoring-using-grafana-prometheus-and-loki), update Loki and Promtail by following the latest respective installation instructions ([Loki](https://docs.codacy.com/v12.0/chart/configuration/monitoring/#2-installing-loki), [Promtail](https://docs.codacy.com/v12.0/chart/configuration/monitoring/#3-installing-promtail)).
```

>     1.  If using the monitoring functionality [Grafana + Prometheus + Loki](https://docs.codacy.com/v12.    0/chart/configuration/monitoring/#setting-up-monitoring-using-grafana-prometheus-and-loki), update Loki and Promtail by following the     latest respective installation instructions ([Loki](https://docs.codacy.com/v12.0/chart/configuration/monitoring/#2-installing-loki), [Promtail](https://docs.codacy.com/v12.0/chart/configuration/monitoring/#3-installing-promtail)).

Reviewer 2: Thanks @lolgab. All the links pointing to `/v12.0` are broken until we publish. This is planned and expected, they will work once we publish the relevant SH docs 👍

    1.  Upgrade Kubernetes to v1.25

## Product enhancements

> -   Added support for [Kubernetes 1.25.\*](https://docs.codacy.com/v12.0/chart/requirements/#kubernetes-or-microk8s-cluster-setup) (REL-1306)

Reviewer 3: Should we add another line mentioning adding support for kubernetes 1.24 or we'll consider it implied since we have support for 1.25? Checked the release notes for sh 11 and we only mentioned support for 1.23 in those.

Reviewer 4: Fine, we'll proceed with option 2.

Reviewer 4: @tfrgoncalves We already added support to K8s 1.24 on [SH v12 release notes (current PR)](https://feat-release-notes-self-hosted-12-0--docs-codacy.netlify.app/release-notes/self-hosted/self-hosted-v12.0.0/#product-enhancements), and removed K8s 1.24 from [SH v11 requirements](https://docs.codacy.com/v11.0/chart/requirements/#kubernetes-or-microk8s-cluster-setup).

Can you please validate and give us approval?

📢 [Visit the Codacy roadmap](https://roadmap.codacy.com) and <span class="skip-vale">let us know</span> your feedback on both new and planned product updates!

## Upgrading Codacy Self-hosted

Follow the steps below to upgrade to Codacy Self-hosted v12.0.0:

1.  Check the [release notes for all Codacy Self-hosted versions](../index.md#self-hosted) **between your current version and the most recent version** for breaking changes and follow the instructions provided <span class="skip-vale">carefully</span>.

1.  Follow the instructions to [upgrade your Codacy Self-hosted instance](https://docs.codacy.com/v12.0/chart/maintenance/upgrade/).

1.  Update your Codacy command-line tools to the following versions:

    -   [Codacy Analysis CLI 7.8.3](https://github.com/codacy/codacy-analysis-cli/releases/tag/7.8.3)
    -   [Codacy Coverage Reporter 13.13.1](https://github.com/codacy/codacy-coverage-reporter/releases/tag/13.13.1)

## Breaking changes

-   If you're updating Kubernetes to v1.25, upgrade to Codacy Self-hosted v12.0.0 before doing so, as some Kubernetes endpoints have been deprecated and may lead to service outage if used with earlier versions of Codacy Self-hosted.

> -   <!-- TODO memory requirements https://codacy.slack.com/archives/CSSSHAK9N/p1688643146512439?thread_ts=1688632097.010499&cid=CSSSHAK9N -->
> 

Author: ```suggestion
```

Reviewer 2: 👍 Removed.

---

# Self-hosted v12.0.0

These release notes are for [Codacy Self-hosted v12.0.0](https://github.com/codacy/chart/releases/tag/12.0.0){: target="_blank"}, released on June 21, 2023.<!-- TODO Update release date -->

📢 [Visit the Codacy roadmap](https://roadmap.codacy.com) and <span class="skip-vale">let us know</span> your feedback on both new and planned product updates!

## Upgrading Codacy Self-hosted

Follow the steps below to upgrade to Codacy Self-hosted v12.0.0:

1.  Check the [release notes for all Codacy Self-hosted versions](../index.md#self-hosted) **between your current version and the most recent version** for breaking changes and follow the instructions provided <span class="skip-vale">carefully</span>.

1.  Follow the instructions to [upgrade your Codacy Self-hosted instance](https://docs.codacy.com/v12.0/chart/maintenance/upgrade/).

1.  Update your Codacy command-line tools to the following versions:

    -   [Codacy Analysis CLI 7.8.3](https://github.com/codacy/codacy-analysis-cli/releases/tag/7.8.3)
    -   [Codacy Coverage Reporter 13.13.1](https://github.com/codacy/codacy-coverage-reporter/releases/tag/13.13.1)

> ## Breaking changes

Reviewer 2: @tfrgoncalves 

I reordered the breaking changes section to clarify the correct order of instructions.

* Can you double-check that the order is correct?
* I realized that we're not mentioning anything about v1.24. Do any of the steps apply if the user updates to v1.24? If so, could you let me know which ones?

Thanks!

cc @lolgab @prcr

Reviewer: OK, so this means we need to decouple the deprecated `scheduler.*` feature from the upgrade to Kubernetes 1.25. :+1:

Reviewer 2: @tfrgoncalves I decoupled the `scheduler.*` instructions from the rest of the upgrade instructions. I'm awaiting your confirmation 👍

Reviewer 3: Making sense to me 👍 Awesome work!

    Assuming you're using the default `codacy` namespace:

    1.  Remove all `scheduler.*` settings from `values.yaml`

    1.  Update the ConfigMap of `worker-manager`:

        ```bash
        kubectl patch configmap/worker-manager --namespace codacy --type merge --patch '{"data":{"CONFIG_FORCE_codacy_kubernetes_scheduler_enable": "false","CONFIG_FORCE_codacy_kubernetes_scheduler_name":"default-scheduler"}}'
        ```

    1.  Restart the `worker-manager` Deployment:

        ```bash
        kubectl rollout restart deployment/worker-manager -n codacy
        ```

    1.  List any pending `worker` Pods:

        ```bash
        kubectl get pods -l app=worker --field-selector=status.phase=Pending -n codacy
        ```

    1.  If there are any pending `worker` Pods, remove them:

        !!! warning
            This is a destructive action. Make sure you understand the consequences.

        ```bash
        kubectl delete pods -l app=worker --field-selector=status.phase=Pending -n codacy
        ```

## Product enhancements

-   Added support for [Kubernetes 1.25.\*](https://docs.codacy.com/v12.0/chart/requirements/#kubernetes-or-microk8s-cluster-setup) (REL-1306)

> -   Added support for [<span class="skip-vale">cgroup</span> v2](https://kubernetes.io/docs/concepts/architecture/cgroups/). (<!-- TODO is there a task tracking this activity?-->)

Reviewer 2: @lolgab I'm having trouble finding the relevant task. Was this done as part of another task, or maybe not tracked on Jira?

Thanks! 🙏 

```suggestion
-   Added support for [<span class="skip-vale">cgroup</span> v2](https://kubernetes.io/docs/concepts/architecture/cgroups/). (<!-- TODO is there a task tracking this activity?-->)
```

Reviewer: [After confirming](https://github.com/codacy/docs/pull/1767#issuecomment-1630853902) with @lolgab, we can actually skip adding this item to the release notes:

```suggestion
```

This version of Codacy Self-hosted includes the tool versions below. The tools that were updated on this version are highlighted in bold:

-   **[Ameba 1.4.3](https://github.com/crystal-ameba/ameba/releases/tag/v1.4.3) (updated from 0.13.1)**
-   Bandit 1.7.0
-   Brakeman 4.3.1
-   bundler-audit 0.6.1
-   **[Checkov 2.3.187](https://github.com/bridgecrewio/checkov/releases/tag/2.3.187) (updated from 2.1.188)**
-   Checkstyle 10.3.1
-   Clang-Tidy 10.0.1
-   CodeNarc 3.2.0
-   CoffeeLint 2.1.0
-   **[Cppcheck 2.10.3](https://github.com/danmar/cppcheck/releases/tag/2.10.3) (updated from 2.10)**
-   Credo 1.4.0
-   CSSLint 1.0.5
-   dartanalyzer 2.17.0
-   **[detekt 1.22.0](https://github.com/detekt/detekt/releases/tag/v1.22.0) (updated from 1.19.0)**
-   **[ESLint 8.44.0](https://github.com/eslint/eslint/releases/tag/v8.44.0) (updated from 8.34.0)**
-   ESLint (deprecated) 7.32.0
-   Faux-Pas 1.7.2
-   Flawfinder 2.0.19
-   **[Gosec 2.15.0](https://github.com/securego/gosec/releases/tag/v2.15.0) (updated from 2.8.1)**
-   Hadolint 1.18.2
-   Jackson Linter 2.14.2
-   JSHint 2.13.5
-   markdownlint 0.26.2
-   PHP Mess Detector 2.13.0
-   PHP_CodeSniffer 3.7.2
-   PMD 6.55.0

> -   **[Prospector 1.10.2](https://github.com/PyCQA/prospector/releases/tag/v1.10.2) (updated from 1.7.7)**

Reviewer: Fixes the broken link:

```suggestion
-   **[Prospector 1.10.2](https://github.com/landscapeio/prospector/releases/tag/1.10.2) (updated from 1.7.7)**
```

Reviewer: I already opened a separate pull request to review the URLs across the other documentation pages: https://github.com/codacy/docs/pull/1789

Jira issues with disabled release notes
-   https://codacy.atlassian.net/browse/TS-369 TODO Notify Paulo that we're ignoring this one: https://codacy.atlassian.net/browse/TS-369?focusedCommentId=56541
-   https://codacy.atlassian.net/browse/TS-356 TODO we may want to notify this: ESlint: install plugin vue-scoped-css
-->

## Upgrading Codacy Self-hosted

Follow the steps below to upgrade to Codacy Self-hosted v12.0.0:

1.  Check the [release notes for all Codacy Self-hosted versions](../index.md#self-hosted) **between your current version and the most recent version** for breaking changes and follow the instructions provided <span class="skip-vale">carefully</span>.

1.  Follow the instructions to [upgrade your Codacy Self-hosted instance](https://docs.codacy.com/v12.0/chart/maintenance/upgrade/).

1.  Update your Codacy command-line tools to the following versions:<!--TODO Update CLI tool versions-->

    -   [Codacy Analysis CLI MAJOR.MINOR.PATCH](https://github.com/codacy/codacy-analysis-cli/releases/tag/MAJOR.MINOR.PATCH)
    -   [Codacy Coverage Reporter MAJOR.MINOR.PATCH](https://github.com/codacy/codacy-coverage-reporter/releases/tag/MAJOR.MINOR.PATCH)

## Breaking changes

> 

Reviewer 3: There are some additional breaking changes to be added here, implemented by @lolgab. Will leave a rough draft for us to work on. The following procedure should be performed by everyone upgrading to v12.0.0 if they have set "scheduler.enable = "true" in values.yaml", even if not performing a Kubernetes version update.
The procedure could be done using kubectl but it could be more dangerous and error prone.

We are deprecating custom scheduling functionality.
If you have set scheduler.enable = "true" in values.yaml, before migrating to Codacy Self-hosted v12.0.0 do:
- Remove scheduler settings from values.yaml
- In worker-manager ConfigMap set  `CONFIG_FORCE_codacy_kubernetes_scheduler_name: default-scheduler `
- Restart worker-manager deployment
- Remove any pending worker Pods
"

Reviewer 2: Thanks! Updated as suggested and moving this discussion: https://github.com/codacy/docs/pull/1767/files#r1239917734

> If you're updating Kubernetes to v1.25, upgrade to Codacy Self-hosted v12.0.0 before doing so, as some Kubernetes endpoints have been deprecated and may lead to service outage if used with Codacy Self-hosted v11.0.0.

Reviewer 2: ```suggestion
If you're updating Kubernetes to v1.25, upgrade to Codacy Self-hosted v12.0.0 before doing so, as some Kubernetes endpoints have been deprecated and may lead to service outage if used with earlier versions of Codacy Self-hosted.
```
